---
title: "CSPC: Consensus Subtypes in Pancreatic Cancer"
author:
- name: Vandana Sandhu
- name: Heewon Seo
- name: Christopher Eeles
  affiliation:
  - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center, University Health Network, Toronto, Ontario, Canada
  email: christopher.eeles@uhnresearch.ca
- name: Benjamin Haibe-Kains
  affiliation:
  - *pm
  - &mbp Department of Medical Biophysics, University of Toronto, Toronto, Canada
  email: benjamin.haibe.kains@utoronto.ca
date: 2021-02-01
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{"CSPC: Consensus Subtypes in Pancreatic Cancer"}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

# COINCIDE

## Preprocess the data

```{r load_data, eval=FALSE}
library(PDATK)
library(S4Vectors)
library(MultiAssayExperiment)
library(qs)

CSPCmaeRaw <- qread('../data/CSPCmae.qs', nthread=16)

# Subset to only feature shared across all data
CSPCmae <- intersectRows(CSPCmaeRaw)
```

```{r feature_ranking, eval=FALSE}
experiments(CSPCmae) <- endoapply(experiments(CSPCmae), rankFeatures)

top20FeaturesPerCohort <- lapply(experiments(CSPCmae), getTopFeatures, numFeats=20)
top20PctFeaturesOverall <- getTopFeatures(CSPCmae, 
    numFeats=floor(0.2*length(rownames(CSPCmae)[[1]])), na.rm=TRUE)
topFeatures <- unique(c(top20PctFeaturesOverall, unlist(top20FeaturesPerCohort)))
```

```{r normalize_MAE, eval=FALSE}
topFeaturesMAE <- CSPCmae[topFeatures, , ]
normMAE <- normalize(topFeaturesMAE, MARGIN=1, method=c('medianImpute', 'center'))
```

## Create and Train a ConsensusClusterModel

```{r , eval=FALSE}
conClustModel <- ConsensusClusteringModel(normMAE, randomSeed=1987)
trainedConClustModel <- trainModel(conClustModel, reps=250)
classifiedConClustModel <- predictClasses(trainedConClustModel)
```

```{r, eval=FALSE}
normalsMAE <- qread('../data/normalsMAE.qs')
topFeatsNormalMAE <- normalsMAE[topFeatures, ]
preprocNormMAE <- normalize(topFeatsNormalMAE, MARGIN=1, 
  method=c('medianImpute', 'center'))
```

```{r, eval=FALSE}
normalsConClustModel <- ConsensusClusteringModel(preprocNormMAE, 
  randomSeed=1987)
trainedNormalsConClustModel <- trainModel(normalsConClustModel, reps=250)
classifiedNormConClustModel <- predictClasses(trainedNormalsConClustModel)
```

```
validatedConClustModel <- validateModel(model=classifiedConClustModel, 
    valData=classifiedNormConClustModel)
```

# Network Community Search




# Gene Signature Comparison




# Biomarker Comparison




# Single Sample Classifier




# COMPASS Subtyping Waterfall



# PharmacoGx Cell-line Subtyping




# PharmacoGx Cell-line Drug Sensitivity




# PDXE Drug Subtyping



# Cellularity Comparison




# Copy Number Abberation




# Cluster Meta-effect Size




# Pathway Analysis




# Published Classifier Subtyping



# References


