---
title: "CSPC: Consensus Subtypes in Pancreatic Cancer"
author:
- name: Vandana Sandhu
- name: Heewon Seo
- name: Christopher Eeles
  affiliation:
  - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center, University Health Network, Toronto, Ontario, Canada
  email: christopher.eeles@uhnresearch.ca
- name: Benjamin Haibe-Kains
  affiliation:
  - *pm
  - &mbp Department of Medical Biophysics, University of Toronto, Toronto, Canada
  email: benjamin.haibe.kains@utoronto.ca
date: 2021-02-01
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{"CSPC: Consensus Subtypes in Pancreatic Cancer"}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

# COINCIDE

This vignette is incomplete and is not intended for users at this time.

## Preprocess the data

```{r load_data, eval=FALSE}
library(PDATK)
library(S4Vectors)
library(MultiAssayExperiment)
library(qs)
library(ggplot2)

seed <- 1990
reps <- 1000

ICGC_SE_list <- qread('vignettes/ICGC_SEs.qs', nthread=16)
names(ICGC_SE_list) <- gsub('-', '_', names(ICGC_SE_list))
for (i in seq_along(ICGC_SE_list)) {
    assayNames(ICGC_SE_list[[i]]) <- gsub('^.*-', '', assayNames(ICGC_SE_list[[i]]))
    assays(ICGC_SE_list[[i]]) <- rev(assays(ICGC_SE_list[[i]]))
}
CSPCmaeRaw <- MultiAssayExperiment(ICGC_SE_list)

# Subset to only feature shared across all data
CSPCmae <- intersectRows(CSPCmaeRaw)
```

```{r feature_ranking, eval=FALSE}
experiments(CSPCmae) <- endoapply(experiments(CSPCmae), rankFeatures)

top20FeaturesPerCohort <- lapply(experiments(CSPCmae), getTopFeatures, numFeats=20)
top20PctFeaturesOverall <- getTopFeatures(CSPCmae, 
    numFeats=floor(0.2*length(rownames(CSPCmae)[[1]])), na.rm=TRUE)
topFeatures <- unique(c(top20PctFeaturesOverall, unlist(top20FeaturesPerCohort)))
```

```{r normalize_MAE, eval=FALSE}
topFeaturesMAE <- CSPCmae[topFeatures, , ]
normMAE <- PDATK::normalize(topFeaturesMAE, MARGIN=1, 
    method=c('medianImpute', 'scale', 'center'))
```

## Create and Train a ConsensusClusterModel

```{r, eval=FALSE}
set.seed(seed)
conClustModel <- ConsensusMetaclusteringModel(normMAE, randomSeed=seed)
trainedConClustModel <- trainModel(conClustModel, reps=reps)
classifiedConClustModel <- predictClasses(trainedConClustModel)
```

<!-- ```{r, eval=FALSE}
normalsMAE <- qread('../data/normalsMAE.qs')
topFeatsNormalMAE <- normalsMAE[topFeatures, ]
preprocNormMAE <- normalize(topFeatsNormalMAE, MARGIN=1,
  method=c('medianImpute', 'scale', 'center'))
```

```{r, eval=FALSE}
normalsConClustModel <- ConsensusMetaclusteringModel(preprocNormMAE, 
  randomSeed=seed)
trainedNormalsConClustModel <- trainModel(normalsConClustModel, reps=reps)
classifiedNormConClustModel <- predictClasses(trainedNormalsConClustModel)
``` -->

```{r eval=FALSE}
validatedConClustModel <- validateModel(model=classifiedConClustModel, 
    valData=ConMetaclustModel(trainData=MultiAssayExperiment(), randomSeed=seed))
qsave(validatedConClustModel, file='valConClustModel_seed99791_250reps.qs')
```

# Network Community Search

```{r eval=FALSE}
validatedConClustModel <- qread('valConClustModel_1000r.qs')
NCSmodel <- NetworkCommunitySearchModel(model=validatedConClustModel)
trainedNCSmodel <- trainModel(NCSmodel, p_value = 0)
classifiedNCSmodel <- predictClasses(trainedNCSmodel)
```

```{r eval=FALSE}
clusterNetworkGraph <- plotNetworkGraph(classifiedNCSmodel)
ggsave(clusterNetworkGraph, file='metacluster_network_graph.pdf',
    width=12, height=12)
```

# Metacluster Survival Analysis

```{r eval=FALSE}
metaclusteredMAE <- trainData(classifiedNCSmodel)
SEhasSurvivalColumns <- lapply(experiments(metaclusteredMAE), hasColDataColumns, 
    values=c('vital_status', 'days_to_death'))
mclustSurvMAE <- metaclusteredMAE[,, unlist(SEhasSurvivalColumns)]
experiments(mclustSurvMAE) <- endoapply(experiments(mclustSurvMAE), 
    FUN=SurvivalExperiment, survival_time='days_to_death', 
    event_occurred='vital_status')
experiments(mclustSurvMAE) <- endoapply(experiments(mclustSurvMAE), 
    dropNotCensored)
```

```{r eval=FALSE}
coxModel <- CoxModel(mclustSurvMAE, survivalPredictor='cluster_label')
fitCoxModel <- trainModel(coxModel)
```

```{r eval=FALSE}
survPlots <- plotSurvivalCurves(fitCoxModel, pval=TRUE, pval.coord=c(0.9, 3000))
ggsave(survPlots, file='survPlots.pdf', height=14, width=14)
```

# Gene Signature Comparison




# Biomarker Comparison




# Single Sample Classifier




# COMPASS Subtyping Waterfall



# PharmacoGx Cell-line Subtyping




# PharmacoGx Cell-line Drug Sensitivity




# PDXE Drug Subtyping



# Cellularity Comparison




# Copy Number Abberation




# Cluster Meta-effect Size




# Pathway Analysis




# Published Classifier Subtyping



# References


